File: 220.contains-duplicate-iii.java
/*
 * @lc app=leetcode id=220 lang=java
 *
 * [220] Contains Duplicate III
 */
class Solution {
    public boolean containsNearbyAlmostDuplicate(int[] nums, int k, int t) {
        if (nums == null) return false;
        TreeSet<Long> ts = new TreeSet<>();
        for (int i = 0; i < nums.length; ++i) {
            if (i >= k+1) ts.remove(new Long(nums[i-k-1]));
            Long ceilingOfLowerBound = ts.ceiling(new Long(nums[i]) - new Long(t));
            if (ceilingOfLowerBound != null && 
                ceilingOfLowerBound <= new Long(nums[i]) + new Long(t)) {
                return true;
            }
            ts.add(new Long(nums[i]));
        }
        return false;
    }
}

File: go/220.contains-duplicate-iii.go
/*
 * @lc app=leetcode id=220 lang=golang
 *
 * [220] Contains Duplicate III
 */

// @lc code=start

// TODO: should use Tree set
func containsNearbyAlmostDuplicate(nums []int, k int, t int) bool {
	cache := make(map[int]struct{})
	for i := 0; i <= k && i < len(nums); i++ {
		if appendToCache(cache, nums[i], t) {
			return true
		}
	}
	for i := k+1; i < len(nums); i++ {
		delete(cache, nums[i-k-1])
		if appendToCache(cache, nums[i], t) {
			return true
		}
	}
	return false
}

// return true if exist
func appendToCache(cache map[int]struct{}, v, t int) bool {
	for v2 := range cache {
		if isAcceptedDiff(v, v2, t) {
			return true
		}
	}
	cache[v] = struct{}{}
	return false
}

func isAcceptedDiff(v1, v2, t int) bool {
	if v1 < v2 {
		v1, v2 = v2, v1
	}
	return v1 - v2 <= t
}


// @lc code=end

